### PR TITLE
Add expiration to rrdata()

### PR DESCRIPTION
**Pull request by jablko**
so ensdomains/dnsregistrar#27 can know when proofs expire and reject them.

### CODE REVIEW

**Comment by Arachnid** [comment]
I think this depends on the discussion in #67.

File: contracts/DNSSECImpl.sol [hunk 1]
```diff
@@ -268,38 +268,41 @@ contract DNSSECImpl is DNSSEC, Owned {
...
      * @param name The name to query, in DNS label-sequence format.
-     * @return inception The unix timestamp at which the signature for this RRSET was created.
-     * @return inserted The unix timestamp at which this RRSET was inserted into the oracle.
      * @return hash The hash of the RRset that was inserted.
+     * @return inception The unix timestamp at which the signature for this RRSET was created.
      */
-    function rrdata(uint16 dnstype, bytes calldata name) external view returns (uint32, uint64, bytes20) {
+    function rrdata(uint16 dnstype, bytes calldata name) external view returns (bytes20, uint32, uint32) {
         RRSet storage result = rrsets[keccak256(name)][dnstype];
-        return (result.inception, result.inserted, result.hash);
```
Comment (Arachnid) [inline contracts/DNSSECImpl.sol]: Could you preserve the original order here, please?

Also ,changing the signature here breaks other stuff - would you mind bumping the version number in package.json, and submitting a PR to ensdomains/dnsregistrar to update it to the new interface?

File: contracts/DNSSECImpl.sol [hunk 2]
```diff
@@ -268,38 +268,41 @@ contract DNSSECImpl is DNSSEC, Owned {
...
             require(int32(inception - set.inception) >= 0);
         }
 
-        if (set.hash == keccak256(rrs)) {
-            // Already inserted!
-            return rrs;
+        // Expire no later than the proof
+        bytes memory signersName = input.readName(RRSIG_SIGNER_NAME);
+        uint16 proofType = proof.readUint16(proof.nameLength(0));
+        uint32 proofExpiration = rrsets[keccak256(signersName)][proofType].expiration;
```
Comment (Arachnid) [inline contracts/DNSSECImpl.sol]: Why are we checking the expiration on the RRSIG/DS here? Shouldn't we be using the expiration on our own signature?

The RFC says:

>  If the resolver accepts the RRset as authentic, the validator MUST
>   set the TTL of the RRSIG RR and each RR in the authenticated RRset to
>   a value no greater than the minimum of:
>
>   o  the RRset's TTL as received in the response;
>
>   o  the RRSIG RR's TTL as received in the response;
>
>   o  the value in the RRSIG RR's Original TTL field; and
>
>   o  the difference of the RRSIG RR's Signature Expiration time and the
>      current time.

That only requires us to check the record being inserted and its signature, not the signer's records.

File: contracts/DNSSECImpl.sol (hunk 1 above)
Comment (jablko) [reply contracts/DNSSECImpl.sol]: > Also ,changing the signature here breaks other stuff - would you mind bumping the version number in package.json, and submitting a PR to ensdomains/dnsregistrar to update it to the new interface?

You bet: ensdomains/dnsregistrar#27 and ensdomains/dnsprovejs#20 are the corresponding pull requests. I'll bump the version number in package.json.

> Could you preserve the original order here, please?

Since we're changing the interface (`inserted` -> `expiration`) I'd like to make the new interface nicer for destructuring assignment: Clients (dnsregistrar, dnsprovejs) commonly need to check the hash and expiration, however you only need the inception if you're fast-failing an update, and only when performing an update, not a lookup.

File: contracts/DNSSECImpl.sol (hunk 2 above)
Comment (jablko) [reply contracts/DNSSECImpl.sol]: To perform validation you need a valid DNSKEY and the DNSKEY isn't valid after its signature expires. So if you want to validate the `subdomain.xyz.` DS record, you need the `subdomain.xyz.` DS RRSIG, the `xyz.` DNSKEY, and the `xyz.` DNSKEY RRSIG.

You probably shouldn't create signatures that expire after the DNSKEY ('s signature) itself, but you can resign the DNSKEY without replacing the DNSKEY, so you can imagine: You get the `xyz.` DNSKEY and signature and submit them to the oracle where they're validated. Time goes by, the DNSKEY RRSIG expires, a new DNSKEY RRSIG is created, and at this point the `subdomain.xyz.` DS record is signed. The oracle must reject the `subdomain.xyz.` authentication chain unless it includes the new, unexpired `xyz.` DNSKEY RRSIG.

Similarly if you want to validate the `subdomain.xyz.` DNSKEY, you need (minimally) the:
1. `subdomain.xyz.` DNSKEY RRSIG
2. `subdomain.xyz.` DNSKEY (validates the `subdomain.xyz.` DNSKEY RRSIG)
3. `subdomain.xyz.` DS (validates the `subdomain.xyz.` DNSKEY)
4. `subdomain.xyz.` DS RRSIG (validates the `subdomain.xyz.` DS)
5. `xyz.` DNSKEY (validates the `subdomain.xyz.` DS RRSIG)
6. `xyz.` DNSKEY RRSIG (validates the `xyz.` DNSKEY)

and the oracle must reject the authentication chain unless each of the three signatures in this example is valid (not expired).

(whether you're validating the whole chain or just the additional part that it hasn't already validated.)

File: contracts/DNSSECImpl.sol [hunk 3]
```diff
@@ -268,33 +268,41 @@ contract DNSSECImpl is DNSSEC, Owned {
      * @dev Returns data about the RRs (if any) known to this oracle with the provided type and name.
      * @param dnstype The DNS record type to query.
      * @param name The name to query, in DNS label-sequence format.
-     * @return inception The unix timestamp at which the signature for this RRSET was created.
-     * @return inserted The unix timestamp at which this RRSET was inserted into the oracle.
```
Comment (Arachnid) [inline contracts/DNSSECImpl.sol]: You need to add a line here for the `expires` field.

File: contracts/DNSSECImpl.sol (hunk 2 above)
Comment (Arachnid) [reply contracts/DNSSECImpl.sol]: Fair enough on those points.

The RFC *doesn't* require us to shorten the expiration of the newly inserted record to the expiration of the record we're using as proof, though. As far as I can tell, it's valid for a DNSKEY whose own RRSIG expires tomorrow to be used to generate an RRSIG on another record that expires in a month.

If we do enforce this, then no record in the system can live beyond the next root key rollover, or past the expiration time of its TLDs signatures.

File: test/TestDNSSEC.js [hunk 4]
```diff
@@ -160,21 +184,21 @@ contract('DNSSEC', function(accounts) {
         type: 'DNSKEY',
         class: 'IN',
         ttl: 3600,
-        data: { flags: 0x0101, algorithm: 253, key: Buffer.from('0000', 'HEX') }
```
Comment (Arachnid) [inline test/TestDNSSEC.js]: Why the change in flags?

File: contracts/DNSSECImpl.sol (hunk 3 above)
Comment (jablko) [reply contracts/DNSSECImpl.sol]: :+1: Done.

File: test/TestDNSSEC.js (hunk 4 above)
Comment (jablko) [reply test/TestDNSSEC.js]: They aren't actually changed but the put in canonical order to match [`rawSignatureData()`](https://github.com/ensdomains/dnsprovejs/blob/master/lib/dns/result.js#L53).

File: contracts/DNSSECImpl.sol (hunk 2 above)
Comment (jablko) [reply contracts/DNSSECImpl.sol]: > As far as I can tell, it's valid for a DNSKEY whose own RRSIG expires tomorrow to be used to generate an RRSIG on another record that expires in a month.

I think we agree that if the `xyz.` DNSKEY RRSIG expires tomorrow and it's used to generate a `subdomain.xyz.` DS RRSIG that expires in a month, then validating the `subdomain.xyz.` DS authentication chain will fail after tomorrow, as will validating the `xyz.` DNSKEY (after tomorrow, and assuming there are no additional signatures).

However validating the `xyz.` DNSKEY today will succeed,  so just to restate your point, if you validate the `xyz.` DNSKEY today, then you can validate the `subdomain.xyz.` DS record after tomorrow (after the `xyz.` DNSKEY RRSIG expires).

The RFC includes plenty of language that says 1) validating an RRset (e.g. the `subdomain.xyz.` DS RRset) requires a valid RRSIG, 2) a valid RRSIG is not expired and requires a valid DNSKEY, and 3) validating a DNSKEY requires a valid RRSIG ... Implicit is that the DNSKEY is valid *at the time you validate the RRset*. It certainly doesn't say the opposite: That you can validate the DNSKEY and validate a dependent authentication chain at completely different times.

(although you have enough material to validate the DNSKEY at any point before its existing signatures expire.)

If it did allow you to validate them at completely different times, then a security-aware resolver could legitimately return an RRset that you couldn't validate (because its authentication chain depended on an expired RRSIG) and my understanding is that's not permitted.

To my eye,
> The RRSIG records that accompany zone data have defined inception and expiration times that establish a validity period for the signatures and the zone data the signatures cover.
> [RFC 4033 Section 8 Zone Considerations](https://tools.ietf.org/html/rfc4033#section-8)

comes close to explicitly stating that a DNSKEY (the zone data the signatures cover) is not valid after the signature that covers it expires, and
> The validator first checks the RRSIG RR to verify that it covers the RRset, has a valid time interval, and identifies a valid DNSKEY RR.
> [RFC 4035 Section 5.3 Authenticating an RRset with an RRSIG RR](https://tools.ietf.org/html/rfc4035#section-5.3)

that validating an RRset requires a valid (not expired) DNSKEY (DNSKEY RRSIG).

Comment (Arachnid) [reply contracts/DNSSECImpl.sol]: Let me provide a concrete example to see if we're on the same page:

 - The .xyz DNSKEY RRSIG expires tomorrow.
 - It is used to produce an RRSIG on DS subdomain.xyz with an expiration time a month from now.

My understanding from the RFC is this:

 1. We can only use this RRSIG to validate the DS until tomorrow.
 2. If we validate the record before tomorrow, we can store it with an expiration of up to a month from now (depending on other TTLs etc as described in the RFC).
 3. We can therefore use the DS record to validate DNSKEY subdomain.xyz for up to a month from now.

The section I quoted earlier provides a list of criteria constraining the TTL of records, and it *doesn't* specify the TTL of the DNSKEY or DS that was used to validate it. If the intention was for this to happen, it seems to me that it would be added as an additional bullet point to that list.

> If it did allow you to validate them at completely different times, then a security-aware resolver could legitimately return an RRset that you couldn't validate (because its authentication chain depended on an expired RRSIG) and my understanding is that's not permitted.

Bear in mind that 1) DNS is not a globally synchronised database; resolvers are only aware of local conditions at the time of signing, and 2) RRSIGs don't specify the RRSIG they depend on; the resolver can fetch any valid RRSIG from the parent nameserver to continue validation.

Comment (jablko) [reply contracts/DNSSECImpl.sol]: We're on the same page, except that the `subdomain.xyz.` DS can be fresh TTL-wise and expired signature-wise ([RFC 4033 Section 8.1 TTL Values vs. RRSIG Validity Period](https://tools.ietf.org/html/rfc4033#section-8.1)) which would be the case in this example.

If someone submits the `subdomain.xyz.` DNSKEY after the `xyz.` DNSKEY RRSIG expires, [RFC 4035 Section 5.2 Authenticating Referrals](https://tools.ietf.org/html/rfc4035#section-5.2) says:
> Given a DS RR for a delegation [e.g. the `subdomain.xyz.` DS], the child zone's apex DNSKEY RRset [the `subdomain.xyz.` DNSKEY] can be authenticated if all of the following hold:
> - The DS RR has been authenticated using some DNSKEY RR in the parent's apex DNSKEY RRset [the `xyz.` DNSKEY] (see Section 5.3).
> - ...

"The DS RR has been authenticated" when? We agree it can't be authenticated after the `subdomain.xyz.` DS RRSIG expires, that's why we can only store it with an expiration of up to a month from now. Except [RFC 4035 Section 5.3.1 Checking the RRSIG RR Validity](https://tools.ietf.org/html/rfc4035#section-5.3.1) says:
> Note that this authentication process is only meaningful if the validator authenticates the DNSKEY RR before using it to validate signatures.

and the `xyz.` DNSKEY can't be authenticated after the `xyz.` DNSKEY RRSIG expires.

In other words the only reason to store the `subdomain.xyz.` DS with an expiration of up to a month from now is because it can't be authenticated after that, but that means we should only store it with an expiration of tomorrow. Otherwise we can store it indefinitely.

My understanding is, given a collection of RRSIG, DNSKEY, and DS records, there are zero or more authentication chains (paths from an RRset to a trust anchor). Each authentication chain is broken when its first component RRSIG expires, and when no chains remain, the RRset is no longer "valid".

It might still be valid, there might be additional RRSIG, DNSKEY, and DS records outside our collection, new RRSIGs can cover existing DNSKEY/DS records (as you point out), but you can't prove it without additional material (as you point out).

Comment (Arachnid) [reply contracts/DNSSECImpl.sol]: > We're on the same page, except that the subdomain.xyz. DS can be fresh TTL-wise and expired signature-wise (RFC 4033 Section 8.1 TTL Values vs. RRSIG Validity Period) which would be the case in this example.

I don't think this is the case. 8.1, and the other sections we've discussed require limiting the subdomain.xyz DS TTL to expire no later than its RRSIG record - they don't require that it also be constrained by the RRSIG on xyz DNSKEY.

Everything else you're saying is compatible with validating the DS when it's inserted, and doesn't require shortening its time to live. Specifically:

> Note that this authentication process is only meaningful if the validator authenticates the DNSKEY RR before using it to validate signatures.

Says 'before' - as in, it must be validated - not 'immediately before'.

If the intent was for signed records to be constrained to expire no later than their signing keys, wouldn't this list from the RFC include that as a constraint?

> If the resolver accepts the RRset as authentic, the validator MUST
set the TTL of the RRSIG RR and each RR in the authenticated RRset to
a value no greater than the minimum of:
> 
> o the RRset's TTL as received in the response;
> 
> o the RRSIG RR's TTL as received in the response;
> 
> o the value in the RRSIG RR's Original TTL field; and
> 
> o the difference of the RRSIG RR's Signature Expiration time and the
current time.

That list, as far as I can tell, is exhaustive; we shouldn't be adding new items to it.

And again - doing this means that every zone is constrained to be re-signed at least as frequently as its parent zone. That doesn't seem like the intent of the system, to me.

> My understanding is, given a collection of RRSIG, DNSKEY, and DS records, there are zero or more authentication chains (paths from an RRset to a trust anchor). Each authentication chain is broken when its first component RRSIG expires, and when no chains remain, the RRset is no longer "valid".

Comment (jablko) [reply contracts/DNSSECImpl.sol]: > Says 'before' - as in, it must be validated - not 'immediately before'.

That's the question: How long are authentication results good for? I haven't read anything that says that list applies to reusing authentication results (or that authentication results can be reused at all).

The list doesn't set out or guarantee authenticity, it applies to freshness, which isn't necessarily the same thing (as far as I can tell).

> And again - doing this means that every zone is constrained to be re-signed at least as frequently as its parent zone. That doesn't seem like the intent of the system, to me.

Not re-signed, re-authenticated. If the `xyz.` DNSKEY is re-signed, then you can continue to use the `subdomain.xyz.` DS RRSIG for up to a month from now. You just can't assume it was re-signed without proof.

Comment (Arachnid) [reply contracts/DNSSECImpl.sol]: Can we please just agree to enforce the restrictions on lifetime explicitly spelled out in the RFC, as a starting point?

Comment (jablko) [reply contracts/DNSSECImpl.sol]: Definitely: I'll move re-authenticating after the proof expires to a separate pull request.
